source/core/shape.cc

```diff
@@ -363,6 +363,195 @@ Sphere<dim>::set_position(const Point<dim> &position)
 }
 
 
+template <int dim>
+void
+Superquadric<dim>::closest_surface_point(
+  const Point<dim> &p,
+  Point<dim> &      closest_point,
+  const typename DoFHandler<dim>::active_cell_iterator & /*cell_guess*/)
+{
+  auto point_in_string = this->point_to_string(p);
+  auto iterator        = this->closest_point_cache.find(point_in_string);
+  if (iterator == this->closest_point_cache.end())
+    {
+      this->closest_surface_point(p, closest_point);
+
+      Point<dim> copy_closest_point{};
+      for (unsigned int d = 0; d < dim; d++)
+        copy_closest_point[d] = closest_point[d];
+      this->closest_point_cache[point_in_string] = copy_closest_point;
+    }
+  else
+    closest_point = iterator->second;
```

Comment:
When you have found the closest point, you can also directly store the value and gradient in the cache.

Reply:
I don't think I can because the function is const, but I'll check again to make sure